File: packages/cli/tests/cli.e2e.test.ts
import { mkdir, readFile, writeFile } from "node:fs/promises";
import { join } from "node:path";

import { execa } from "execa";
import { afterEach, beforeEach, describe, expect, it } from "vitest";
import packageJson from "../package.json" with { type: "json" };
import { rootCommandRoutes } from "../src/cli/root-commands.ts";
import { cliNodeOptions } from "../src/test/helpers/cli-entry.ts";
import {
  createSyncE2EContext,
  type SyncE2EContext,
} from "../src/test/helpers/e2e-context.ts";
import { stripAnsi } from "../src/test/helpers/sync-fixture.ts";

const runCli = async (
  args: readonly string[],
  options?: Readonly<{
    env?: NodeJS.ProcessEnv;
    reject?: boolean;
  }>,
) => {
  return execa(process.execPath, [...cliNodeOptions, ...args], {
    env: options?.env,
    reject: options?.reject,
  });
};

describe("CLI e2e", () => {
  it("shows the version from the real entrypoint", async () => {
    const result = await runCli(["--version"]);

    expect(result.exitCode).toBe(0);
    expect(result.stdout).toContain(`dotweave/${packageJson.version}`);
    expect(result.stderr).toBe("");
  });

  it("shows root help with the new command surface", async () => {
    const result = await runCli([]);

    expect(result.exitCode).toBe(0);
    expect(result.stderr).toBe("");

    const rootCommandNames = [
      "autocomplete",
      ...Object.keys(rootCommandRoutes),
    ];
    for (const commandName of rootCommandNames) {
      expect(result.stdout).toContain(commandName);
    }

    expect(result.stdout).toContain("Launch a shell in the sync directory");
  });

  it("shows help for cd, track, and profile use commands", async () => {
    const [cdHelp, trackHelp, profileHelp] = await Promise.all([
      runCli(["cd", "--help"]),
      runCli(["track", "--help"]),
      runCli(["profile", "use", "--help"]),
    ]);

    expect(cdHelp.stdout).toContain("USAGE");
    expect(cdHelp.stdout).toContain("Launch a child shell rooted");

    expect(trackHelp.stdout).toContain("USAGE");
    expect(trackHelp.stdout).toContain("--mode");
    expect(trackHelp.stdout).toContain("--profile");
    expect(trackHelp.stdout).toContain("--repo");

    expect(profileHelp.stdout).toContain("Profile name to activate");
  });

  it("rejects removed --verbose flag", async () => {
    const result = await runCli(["pull", "--verbose"], { reject: false });

    expect((result.exitCode ?? 0) & 0xff).toBe(252);
    expect(result.stderr).toContain("No flag registered for --verbose");
  });

  it("returns a non-zero exit code for removed command surfaces", async () => {
    const [addResult, removeResult, modeResult, listResult, dirResult] =
      await Promise.all([
        runCli(["add", "~/.gitconfig"], { reject: false }),
        runCli(["remove", "~/.gitconfig"], { reject: false }),
        runCli(["mode", "secret", "~/.gitconfig"], { reject: false }),
        runCli(["list"], { reject: false }),
        runCli(["dir"], { reject: false }),
      ]);

    expect(addResult.exitCode).not.toBe(0);
    expect(addResult.stderr).toContain("not found");
    expect(removeResult.exitCode).not.toBe(0);
    expect(removeResult.stderr).toContain("not found");
    expect(modeResult.exitCode).not.toBe(0);
    expect(modeResult.stderr).toContain("not found");
    expect(listResult.exitCode).not.toBe(0);
    expect(listResult.stderr).toContain("not found");
    expect(dirResult.exitCode).not.toBe(0);
    expect(dirResult.stderr).toContain("not found");
  });
});

let ctx: SyncE2EContext;

beforeEach(async () => {
  ctx = await createSyncE2EContext();
});

afterEach(async () => {
  await ctx.cleanup();
});

describe("CLI sync cycle e2e", () => {
  it("runs a full init-track-push-pull cycle", async () => {
    const configDir = join(ctx.homeDir, ".config", "myapp");
    const configFile = join(configDir, "config.toml");
    const ageKeys = await ctx.createAgeKeyPair();

    await ctx.writeIdentityFile(ageKeys.identity);
    await mkdir(configDir, { recursive: true });
    await writeFile(configFile, "key = value\n");

    await ctx.runCli(["init"]);
    await ctx.runCli(["track", configDir]);
    await ctx.runCli(["push"]);

    await writeFile(configFile, "key = modified\n");
    await ctx.runCli(["pull", "-y"]);

    const content = await readFile(configFile, "utf8");
    expect(content).toContain("key = value\n");
  });

  it("reports status for an initialized repository", async () => {
    const configDir = join(ctx.homeDir, ".config", "myapp");
    const ageKeys = await ctx.createAgeKeyPair();

    await ctx.writeIdentityFile(ageKeys.identity);
    await mkdir(configDir, { recursive: true });
    await writeFile(join(configDir, "config.toml"), "key = value\n");

    await ctx.runCli(["init"]);
    await ctx.runCli(["track", configDir]);

    const result = await ctx.runCli(["status"]);

    expect(result.exitCode).toBe(0);
    const out = stripAnsi(result.stdout);
    expect(out).toContain("Sync status");
    expect(out).toContain("Push changes");
    expect(out).toContain("Add");
  });

  it("reports missing git during repository import without blaming reachability", async () => {
    const ageKeys = await ctx.createAgeKeyPair();
    const keyFile = await ctx.writeIdentityFile(ageKeys.identity);

    const result = await ctx.runCli(
      ["init", "https://example.invalid/dotfiles.git", "--key-file", keyFile],
      {
        env: { PATH: "", Path: "" },
        reject: false,
      },
    );
    const stderr = stripAnsi(result.stderr);

    expect(result.exitCode).toBe(1);
    expect(stderr).toContain("Git is not installed or not on PATH.");
    expect(stderr).toContain(
      "Install Git and ensure the git executable is available on PATH",
    );
    expect(stderr).not.toContain("repository source is reachable");
  });
});

File: packages/cli/test/e2e/cli_e2e_test.dart
// Dart port of `packages/cli/tests/cli.e2e.test.ts`.
//
// The suite-local `runCli` (used by the "CLI e2e" group) spawns the compiled
// executable with only the minimal pass-through environment, mirroring the TS
// helper that spawns node against `src/index.ts` without env overrides. The
// file-level beforeEach/afterEach hooks apply to every test in the file, as
// they do in vitest.

@Timeout(Duration(minutes: 3))
library;

import 'dart:io';

import 'package:dotweave/src/cli/root_commands.dart';
import 'package:dotweave/src/util/version.g.dart';
import 'package:path/path.dart' as p;
import 'package:test/test.dart';

import '../helpers/e2e_context.dart';
import '../helpers/sync_fixture.dart';

Future<CliRunResult> _runCli(
  List<String> args, {
  Map<String, String>? env,
  bool reject = true,
}) async {
  final result = await runCompiledCli(args, env: env);

  if (reject && result.exitCode != 0) {
    throw CliRunException(args, result);
  }

  return result;
}

void main() {
  group('CLI e2e', () {
    test('shows the version from the real entrypoint', () async {
      final result = await _runCli(['--version']);

      expect(result.exitCode, 0);
      expect(result.stdout, contains('dotweave/$packageVersion'));
      expect(result.stderr, '');
    });

    test('shows root help with the new command surface', () async {
      final result = await _runCli([]);

      expect(result.exitCode, 0);
      expect(result.stderr, '');

      final rootCommandNames = ['autocomplete', ...rootCommandRoutes.keys];
      for (final commandName in rootCommandNames) {
        expect(result.stdout, contains(commandName));
      }

      expect(result.stdout, contains('Launch a shell in the sync directory'));
    });

    test('shows help for cd, track, and profile use commands', () async {
      final results = await Future.wait([
        _runCli(['cd', '--help']),
        _runCli(['track', '--help']),
        _runCli(['profile', 'use', '--help']),
      ]);
      final cdHelp = results[0];
      final trackHelp = results[1];
      final profileHelp = results[2];

      expect(cdHelp.stdout, contains('USAGE'));
      expect(cdHelp.stdout, contains('Launch a child shell rooted'));

      expect(trackHelp.stdout, contains('USAGE'));
      expect(trackHelp.stdout, contains('--mode'));
      expect(trackHelp.stdout, contains('--profile'));
      expect(trackHelp.stdout, contains('--repo'));

      expect(profileHelp.stdout, contains('Profile name to activate'));
    });

    test('rejects removed --verbose flag', () async {
      final result = await _runCli(['pull', '--verbose'], reject: false);

      expect(result.exitCode & 0xff, 252);
      expect(result.stderr, contains('No flag registered for --verbose'));
    });

    test('returns a non-zero exit code for removed command surfaces', () async {
      final results = await Future.wait([
        _runCli(['add', '~/.gitconfig'], reject: false),
        _runCli(['remove', '~/.gitconfig'], reject: false),
        _runCli(['mode', 'secret', '~/.gitconfig'], reject: false),
        _runCli(['list'], reject: false),
        _runCli(['dir'], reject: false),
      ]);
      final addResult = results[0];
      final removeResult = results[1];
      final modeResult = results[2];
      final listResult = results[3];
      final dirResult = results[4];

      expect(addResult.exitCode, isNot(0));
      expect(addResult.stderr, contains('not found'));
      expect(removeResult.exitCode, isNot(0));
      expect(removeResult.stderr, contains('not found'));
      expect(modeResult.exitCode, isNot(0));
      expect(modeResult.stderr, contains('not found'));
      expect(listResult.exitCode, isNot(0));
      expect(listResult.stderr, contains('not found'));
      expect(dirResult.exitCode, isNot(0));
      expect(dirResult.stderr, contains('not found'));
    });
  });

  late SyncE2EContext ctx;

  setUp(() async {
    ctx = await createSyncE2EContext();
  });

  tearDown(() async {
    await ctx.cleanup();
  });

  group('CLI sync cycle e2e', () {
    test('runs a full init-track-push-pull cycle', () async {
      final configDir = p.join(ctx.homeDir, '.config', 'myapp');
      final configFile = p.join(configDir, 'config.toml');
      final ageKeys = await ctx.createAgeKeyPair();

      await ctx.writeIdentityFile(ageKeys.identity);
      await Directory(configDir).create(recursive: true);
      await File(configFile).writeAsString('key = value\n');

      await ctx.runCli(['init']);
      await ctx.runCli(['track', configDir]);
      await ctx.runCli(['push']);

      await File(configFile).writeAsString('key = modified\n');
      await ctx.runCli(['pull', '-y']);

      final content = await File(configFile).readAsString();
      expect(content, contains('key = value\n'));
    });

    test('reports status for an initialized repository', () async {
      final configDir = p.join(ctx.homeDir, '.config', 'myapp');
      final ageKeys = await ctx.createAgeKeyPair();

      await ctx.writeIdentityFile(ageKeys.identity);
      await Directory(configDir).create(recursive: true);
      await File(
        p.join(configDir, 'config.toml'),
      ).writeAsString('key = value\n');

      await ctx.runCli(['init']);
      await ctx.runCli(['track', configDir]);

      final result = await ctx.runCli(['status']);

      expect(result.exitCode, 0);
      final out = stripAnsi(result.stdout);
      expect(out, contains('Sync status'));
      expect(out, contains('Push changes'));
      expect(out, contains('Add'));
    });

    test('reports missing git during repository import without blaming '
        'reachability', () async {
      final ageKeys = await ctx.createAgeKeyPair();
      final keyFile = await ctx.writeIdentityFile(ageKeys.identity);

      final result = await ctx.runCli(
        ['init', 'https://example.invalid/dotfiles.git', '--key-file', keyFile],
        env: {'PATH': '', 'Path': ''},
        reject: false,
      );
      final stderr = stripAnsi(result.stderr);

      expect(result.exitCode, 1);
      expect(stderr, contains('Git is not installed or not on PATH.'));
      expect(
        stderr,
        contains(
          'Install Git and ensure the git executable is available on PATH',
        ),
      );
      expect(stderr, isNot(contains('repository source is reachable')));
    });
  });
}
